src/System.Text.ValueBuilder/System/Text/ValueStringBuilder.Format.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Text
+{
+    public ref partial struct ValueStringBuilder
+    {
+        // Undocumented exclusive limits on the range for Argument Hole Index and Argument Hole Alignment.
+        private const int IndexLimit = 1000000; // Note:            0 <= ArgIndex < IndexLimit
+        private const int WidthLimit = 1000000; // Note:  -WidthLimit <  ArgAlign < WidthLimit
+
+        // This is a copy of the StringBuilder.AppendFormatHelper with minor functional tweaks:
+        //
+        //  1. Has a small stackalloc span for formatting Variant value types into.
+        //  2. Doesn't work with ISpanFormattable (the interface is currently internal).
+        //  3. Uses Variant to format with no allocations for value types.
+        //  4. Takes FormatString instead of string for format.
+        //  5. Takes ReadOnlySpan<Variant> instead of ParamsArray.
+        //  6. Code formatting is scrubbed a bit for clarity.
+
+        // Note that Argument Hole parsing can be factored into a helper, perhaps taking a callback
+        // delegate with (int index, int width, ReadOnlySpan<char> itemFormat) or something along those
+        // lines. This would, of course, make the code a little slower, but the advantage of having
+        // shareable logic (between StringBuilder, ValueStringBuilder, etc.) may be worth it.
+
+        public unsafe void Append(FormatString formatString, ReadOnlySpan<Variant> args, IFormatProvider provider = null)
+        {
+            ReadOnlySpan<char> format = formatString.Format;
+
+            int position = 0;
+            int formatLength = format.Length;
+            char current = '\x0';
+            ValueStringBuilder unescapedItemFormat = default;
+
+            // Can't do an inline stackalloc as we can't express that we won't capture the span.
+            char* c = stackalloc char[32];
+            Span<char> initialBuffer = new Span<char>(c, 32);
+            ValueStringBuilder formatBuilder = new ValueStringBuilder(initialBuffer);
+
+            ICustomFormatter customFormatter = null;
+            if (provider != null)
+            {
+                customFormatter = (ICustomFormatter)provider.GetFormat(typeof(ICustomFormatter));
+            }
+
+            while (true)
+            {
+                // Scan for an argument hole (braces)
+                while (position < formatLength)
```

Comment:
a core concern i have here is that it seems like (and correct me if i'm wrong), everyone needs to know how to process a format string.  That seems highly undesirable and error prone.  Is there a canonical, lightweight (ideally, stack allocated) parser that one can use to drive any person who needs this?

Alternatively, instead of having to parse this out at runtime, would it be perhaps better to have the C# compiler automatically provide a prebaked struct that defines all this?  For example, in JS/TS when you use interpolated strings, you get *3* arrays back from the compiler:

1. the array of literals, with escapes interpreted.  i.e. if the original code was: ``` foo `bar \t { baz } quux` ```, then this woudl have `["bar \t ", " quux"]`.  
2. the array of literals, with escapes untouched.  Using hte above example, you would get back: `["bar \\t ", " quux"]`.  
3. the values of the interpolations.  i.e. whatever value 'baz' has.

It's then easy for consumers to understand what they have and stitch things together.  There's no need for them to figure out htings like "oh, i have a `{`, is it also followed by a `{`?".  The compiler has already figured that out for them.

Reply:
Note: by stack allocated parser, i mean one similar to how you might see an event driven json/xml parser.  so you would do this:

```c#
var parser = new Parser(formatString);
while (parser.MoveToNext())
{
    switch (parser.Current)
    {
         case Text:
         case FormattingParameter:
    }
}
```

I'd prefer this all just be in a stack allocated ref-structure that teh compiler could just pass into the method.  But, absent that, having a parser that anyone can use seems preferable.

Reply:
Note that this has to be a drop-in replacement for `StringBuilder.AppendFormat` so we're constrained somewhat.

I'm not sure I follow on why everyone needs to know how to parse the format string. There currently is only one parser in .NET string formatting, this is simply a low-allocation version of that. It isn't intended to be customized / reimplemented.

Having the compiler provide a pre-parsed format string would require having a constant format string. That isn't always the case. While we could support some sort of pre-parsed data structure I'm skeptical that we'd be able to do so in a way that is more performant than walking through the string. Additionally it is currently impossible to format arguments to prep this without allocating a lot of string objects.

While making this implementation pluggable would be nice, it would likely be difficult to get close to the performance of a monolithic method. That said I'd love to see deeper dives into this sort of thing if anyone is willing to fully explore it.